File: application/backend-credit-card-enrollment/backend-c#/CreditCardEnrollment/CreditCard/Enrollment/Event/EnrollmentDeclined.cs
using CreditCardEnrollment.Common.Event;

namespace CreditCardEnrollment.CreditCard.Enrollment.Event;

public class EnrollmentDeclined : TransformationEvent<Aggregate.Enrollment>
{
    public required string ReasonCode { get; init; }
    public required string ReasonDescription { get; init; }

    public override Aggregate.Enrollment TransformAggregate(Aggregate.Enrollment aggregate)
    {
        return new Aggregate.Enrollment
        {
            AggregateId = AggregateId,
            AggregateVersion = AggregateVersion,
            UserId = aggregate.UserId,
            ProductId = aggregate.ProductId,
            Status = Aggregate.EnrollmentStatus.Declined,
            AnnualIncomeInCents = aggregate.AnnualIncomeInCents,
            EnrollmentFirstRequestedOn = aggregate.EnrollmentFirstRequestedOn
        };
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/event/EnrollmentDeclined.ts
import { TransformationEvent } from '../../../common/event/TransformationEvent';
import { Enrollment } from '../aggregate/Enrollment';
import { EnrollmentStatus } from '../aggregate/EnrollmentStatus';

export class EnrollmentDeclined extends TransformationEvent<Enrollment> {
    constructor(
        eventId: string,
        aggregateId: string,
        aggregateVersion: number,
        correlationId: string,
        causationId: string,
        recordedOn: Date,
        public readonly reasonCode: string,
        public readonly reasonDescription: string
    ) {
        super(eventId, aggregateId, aggregateVersion, correlationId, causationId, recordedOn);
    }

    transformAggregate(aggregate: Enrollment): Enrollment {
        return new Enrollment(
            this.aggregateId,
            this.aggregateVersion,
            aggregate.userId,
            aggregate.productId,
            EnrollmentStatus.Declined,
            aggregate.annualIncomeInCents,
            aggregate.enrollmentFirstRequestedOn
        );
    }
}